MdeModulePkg/Core/Dxe/Mem/Page.c

```diff
@@ -1765,6 +1772,157 @@ CoreFreePages (
+  // interrupt came in and changed the memory map, it would be a bug if it was for this region (since it is being
+  // freed) and the only reason we took the lock now was to parse the list, which we are done with for now)
+  CoreReleaseMemoryLock ();
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  // The free pages code will attempt to zero the freed pages if DebugClearMemory is enabled. However, there is no
+  // guarantee that the memory is writeable, which can cause the core to crash. Furthermore, if DebugClearMemory is
+  // not enabled and NX memory protection is not enabled, we can end up passing pages freed here directly to another
+  // caller in the next allocate pages call, which may not be writeable, which is the expection. So, we need to ensure
+  // the memory is writeable before we attempt to zero it or pass it out. We must do this outside of the memory lock to
+  // avoid deadlocks going through the GCD, which may cause page allocations to occur. It is safe to call the GCD from
+  // here, because the GCD only uses pool backed memory and CoreFreePoolPages() does not call through this path, so we
+  // won't hit a deadlock on the GCD lock. We don't attempt to call this before gCpu is initialized, because before
+  // that point, page attributes can't be set anyway and we may be very early in GCD init and not have the GCD
+  // populated yet.
+  if (gCpu == NULL) {
+    return CoreInternalFreePages (Memory, NumberOfPagesAligned, MemoryType, IsGuarded);
+  }
+
+  CurrentAddress = Memory;
+  EndAddress     = Memory + EFI_PAGES_TO_SIZE (NumberOfPagesAligned);
+  while (CurrentAddress < EndAddress) {
+    Status = CoreGetMemorySpaceDescriptor (
+               CurrentAddress,
+               &Descriptor
+               );
+
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a: failed to get memory space descriptor for %llx with %llx pages\n",
+        __func__,
+        Memory,
+        NumberOfPages
+        ));
+      OperationStatus = Status;
+      ASSERT_EFI_ERROR (Status);
+
+      // if we can't get the descriptor, we don't know what base address to try next, so all we can do is try one
+      // page forward
+      CurrentAddress += EFI_PAGE_SIZE;
+      continue;
+    }
+
+    DescEnd = Descriptor.BaseAddress + Descriptor.Length;
+    // ensure that we only change the attributes for the range that we are interested in, not the entire descriptor,
+    // we may also be in the middle of a descriptor, so ensure our length is not larger than the descriptor length
+    if (EndAddress > DescEnd) {
+      CurrentLength = DescEnd - CurrentAddress;
+    } else {
+      CurrentLength = EndAddress - CurrentAddress;
+    }
+
+    // we always set the attributes here, don't check the descriptor attributes, because they may be out of sync
+    // if the attributes were changed with the CPU arch protocol or the memory attribute protocol
+    Status = CoreSetMemorySpaceAttributes (

```

Comment:
At the end of this, this memory will be left as RWX, right?

Reply:
It depends on what is set there, but RWX is likely. However, inside of CoreInternalFreePages now, the NX policy is applied, so it will get NX applied to it if the platform has that configured. If not, then all memory will be RWX (except whatever drivers change). So at the end of the free call, attributes will be set as the platform has configured.